Removed breaking type check from nagios module

silverwizard opened the pull request:
##### SUMMARY
Python 2.x (unknown point release when it started, I used Python 2.7.15rc1) returns `<type 'int'>` when casting minutes in the nagios module. Python 3.x (also unknown point release, I used Python 3.6.6) returns `<class 'int'>`. Changing from a numeric primitive to a class seems to have caused `types.IntTypes` to throw an exception when it attempted to compare the type, meaning that the two actions could never be used on a system defaulting to python 3.x.

This fix removes the `types.IntType` check, since the cast will throw an exception in any situation where the type check would fail. 

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
Nagios

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
This issue was caused by just running the nagios module with the action of "downtime".
<!--- Paste verbatim command output below, e.g. before and after your change -->
```
(dev):~/devops/ansible ❯ ansible -i ec2.py tag_primary_role_nagios -u xxx -m nagios -a "action=downtime service=host host=localhost" -vvv
ansible 2.6.2
  config file = /root/devops/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.6 (default, Sep 12 2018, 18:26:19) [GCC 8.0.1 20180414 (experimental) [trunk revision 259383]]
Using /root/devops/ansible/ansible.cfg as config file
Parsed /root/devops/ansible/ec2.py inventory source with script plugin
META: ran handlers
Using module file /usr/local/lib/python3.6/dist-packages/ansible/modules/monitoring/nagios.py
<35.168.187.224> ESTABLISH SSH CONNECTION FOR USER: xxx
<35.168.187.224> SSH: EXEC ssh -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=xxx -o ConnectTimeout=40 -o ControlPath=/root/.ansible/cp/%h-%p-%r 35.168.187.224 '/bin/sh -c '"'"'/usr/bin/python && sleep 0'"'"''
<35.168.187.224> (1, b'\n{"msg": "invalid entry for minutes", "failed": true, "exception": "  File \\"/tmp/ansible_od56pxwn/ansible_module_nagios.py\\", line 300, in main\\n    if not isinstance(m, types.IntType):\\n", "invocation": {"module_args": {"action": "downtime", "service": "host", "host": "localhost", "services": "host", "author": "Ansible", "comment": "Scheduling downtime", "minutes": "30", "cmdfile": "/var/lib/nagios3/rw/nagios.cmd", "servicegroup": null, "command": null}}}\n', b'bash: warning: setlocale: LC_ALL: cannot change locale (en_CA.UTF-8)\n')
The full traceback is:
  File "/tmp/ansible_od56pxwn/ansible_module_nagios.py", line 300, in main
    if not isinstance(m, types.IntType):

nagios | FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "action": "downtime",
            "author": "Ansible",
            "cmdfile": "/var/lib/nagios3/rw/nagios.cmd",
            "command": null,
            "comment": "Scheduling downtime",
            "host": "localhost",
            "minutes": "30",
            "service": "host",
            "servicegroup": null,
            "services": "host"
        }
    },
    "msg": "invalid entry for minutes"
}
(dev):~/devops/ansible ❯ 

```

Commit:
Removed extraneous type check from nagios module, in order to allow python 3.x

Commit:
Removed now useless import types

silverwizard commented:
ready_for_review

felixfontein commented:
It's probably better to replace the check with `six.integer_types`
```.py
from ansible.module_utils.six import integer_type
[...]
if isinstance(..., integer_types):
    [...]
```

felixfontein commented:
Or remove it and rely on `int`; but please change the generic `Exception` to `ValueError`.

felixfontein commented:
Also, you need a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to).

silverwizard commented:
The type is cast to integer the line before, so the exception would be something larger than a type exception. And if something is not an integer after being cast to an integer, the cast should be what throws the exception. The type check only exists to cause problems in the future.

Commit:
Added changelog fragment

felixfontein commented:
Can you give an example when `int(x)` raises something else than a `TypeError`?

felixfontein commented on changelogs/fragments/495680-type-checking.yaml:
```diff
@@ -0,0 +1,2 @@
+---
+- Removed redundant type check in Nagios module which only existed to cause crashes
```
```
bugfixes:
- "nagios - Removed redundant type check which caused crashes."
```

felixfontein replied:
If you fix this, I think the PR is good to merge.

felixfontein commented:
> Can you give an example when `int(x)` raises something else than a `TypeError`?

Looks like I forgot the obvious `ValueError`, for example with `int('1.1')`.

jctanner commented:
!waffling stale_ci

felixfontein committed:
Update changelog.

felixfontein reviewed (approved):
LGTM

Can anyone else take a look, or even verify whether / that this works? @Klaas- maybe? :)

Klaas- commented:
@felixfontein I'll try to get to it this week, I can test it on fedora with ansible 2.8

Klaas- commented:
is there a prior check if int is minutes? if I put in an invalid type -- ie a string I get
```
# ansible -i "server.tld," -m nagios -a "action=downtime minutes=asd service=host host=server.tld" all
server.tld | FAILED! => {
    "changed": false,
    "msg": "argument minutes is of type <type 'str'> and we were unable to convert to int: invalid literal for int() with base 10: 'asd'"
}
```
whereas I would expect it to return "minutes is not a valid integer" from code changes.

This is ansible 2.7 on rhel7 (python 2.7) install though. Nonetheless it still sets downtimes properly when I supply a valid value.

felixfontein commented:
In the meantime (i.e. somewhen after this PR has been started), a type of `int` has been added to `minutes`: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/monitoring/nagios.py#L264 (it was added in #55026)

Therefore, Ansible ensures that it is an integer. The code in this PR does not have this type, though.

Maybe we should rebase this PR and simply remove the checks, since `minutes` already is an integer. @silverwizard do you want to do that, or do you want me to do that for you?

Commit:
Rebased and removed check due to module adding earlier guardrails

Commit:
Updated changelog to mention earlier fix adding now completely removed guardrails

silverwizard commented:
Fixed, I rebased then removed the guardrail and updated changelog. Testing locally seems to be working for me.

Travis is having trouble finding the commit for the changelog commit, and I don't seem to have an option to rerun tests.

felixfontein commented on lib/ansible/modules/monitoring/nagios.py:
```diff
@@ -294,13 +293,7 @@ def main():
-                module.fail_json(msg='minutes must be a number')
-        except Exception:
-            module.fail_json(msg='invalid entry for minutes')
+        m = int(minutes)
```
The value of `m` is never used again. I think you can simply remove this whole conversion.

felixfontein replied:
In line 285, also remove mention of `minutes`.

felixfontein commented on lib/ansible/modules/monitoring/nagios.py:
```diff
@@ -315,12 +308,7 @@ def main():
         if not servicegroup:
             module.fail_json(msg='no servicegroup selected to set downtime for')
         # Make sure minutes is a number
```
Here also remove this comment.

felixfontein reviewed (commented):
I think you can

felixfontein committed:
Remove superfluous type checks. Fix docs type.

felixfontein commented:
After rebasing by `devel`, this should now be fine since `devel` contains `type='int'` in the argspec.

felixfontein commented:
Ok, we're at a point where a rebase is strictly necessary.

felixfontein committed:
Update ignore.txt.

felixfontein commented:
I think this looks good now. If anyone (@Klaas- ? @silverwizard ?) could quickly test this again before I merged, I'd be happy :)

resmo reviewed (approved):
LGTM

felixfontein commented:
Thanks everyone for your work! I'll create a backport to stable-2.8.
